### PR TITLE
contracts: submit pre-signed transaction

### LINKED ISSUE #179: client-js: Encrypt plain transactions in sendTransaction()
In the onchain-signer and demo-voting examples, we use plain, unwrapped provider in order to call `sendTransaction()` of the on-chain generated and signed EIP-155 transaction, because the wrapper aborts with `Error: Un-enveloped data was passed to sendRawTransaction, which is likely incorrect. Is the dapp using the Sapphire compat lib correctly?`.

The downside of this is that **the transaction is not encrypted**, so we need to encrypt the inner call on the chain to get privacy. This is expensive and makes no sense since the client could encrypt it with the symmetric key when submitting it.

Note: encryption is not related to the fact that the tx was signed by someone else (=on-chain signer in our case), so this should be perfectly doable.

### PR DESCRIPTION
**Pull request by CedarMist**
Fixes #179:

    Error: Un-enveloped data was passed to sendRawTransaction, which is likely incorrect. Is the dapp using the Sapphire compat lib correctly?

I have added tests transaction envelope tests:

    ✔ Wrapper encrypts self-signed transaction calldata (4038ms)
    ✔ Other-Signed transaction submission via un-wrapped provider (4223ms)
    ✔ Other-Signed transaction submission via wrapped provider (4115ms)
    ✔ Self-Signed transaction submission via wrapped provider (4061ms)

This slightly modifies sapphire.wrap behavior so if it receives a signed but unencrypted transaction from an address which it can't sign for the transaction will pass through as-is.

### CODE REVIEW

**Comment by CedarMist** [comment]
The `repackRawTx` function rejects things that it could easily handle.

And if you CBOR encode the envelope you end up with the contract receiving CBOR encoded data, rather than the calldata...

**Comment by CedarMist** [comment]
This will fix #179

**Comment by CedarMist** [comment]
There are a few scenarios we want to handle with `sendRawTransaction`:

 * User submits signed but unencrypted transaction
   * Wrap in encrypted envelope, no need to re-sign it, submit to RPC endpoint
 * User submits signed & encrypted transaction
   * Pass it verbatim to RPC endpoint
 * User submits unsigned unencrypted transaction
   * Sign it, encrypt it & submit to RPC endpoint  

However, currently the wrapper will raise an error if you submit a transaction signed by somebody else, it will try to re-sign the transaction.

**Comment by CedarMist** [comment]
So... looks like the error I want to fix is fundamentally unfixable.

But, instead I added a test to make sure the EIP155 module can sign and submit transactions, and that transactions submitted by the wrapped provider are encrypted.

**Comment by CedarMist** [comment]
So I can fix this error where we submit pre-signed transactions through the provider which is attached to a signer with a different address. But we can't automatically encrypt pre-signed transactions.

**Comment by matevz** [comment]
> So I can fix this error where we submit pre-signed transactions through the provider which is attached to a signer with a different address. But we can't automatically encrypt pre-signed transactions.

Why not? Encryption with an ephemeral key derived from the runtime and the client key has nothing to do with who signed the tx.

**Comment by CedarMist** [comment]
> Why not? Encryption with an ephemeral key derived from the runtime and the client key has nothing to do with who signed the tx.

The calldata is signed, modifying the calldata by wrapping it in an encrypted envelope would invalidate the existing signature.

When I use:

```
const [owner] = await ethers.getSigners();
const provider = testContract.provider;
const wallet = sapphire.wrap(owner.connect(provider));
```

The `repackRawTx` function encrypts the calldata then asks the signer to re-sign the modified transaction, however with the JSON-RPC Signer it's 

```
Error: cannot alter JSON-RPC Signer connection (operation="connect", code=UNSUPPORTED_OPERATION, version=providers/5.7.2)
```

If I wrap the `owner` directly, e.g. `const wallet = sapphire.wrap(owner);` I get:

```
Error: signing transactions is unsupported (operation="signTransaction", code=UNSUPPORTED_OPERATION, version=providers/5.7.2)
```

And only via creating a new `ethers.Wallet` directly and `connect()` it to the provider does it work because it can sign the transaction.

If this happens via MetaMask and the dApp signs a transaction without encrypting it then sends it through a wrapped signer it will try to encrypt it before submission.

I have slightly modified the behavior and added test cases so it you pass a transaction not signed by the signers address it will not attempt to encrypt it and pass it on. This makes submitting the gasless transactions easier because you don't need to setup a different unwrapped provider to submit it.

------------------------

**Comment by CedarMist** [comment]
This may be relevant: https://github.com/oasisprotocol/sapphire-paratime/issues/106

File: clients/js/src/compat.ts [hunk 1]
```diff
@@ -573,11 +573,13 @@ async function repackRawTx(
     );
   } catch (e) {
     if (e instanceof EnvelopeError) throw e;
-    if (globalThis?.process?.env?.NODE_ENV !== 'test') {
-      console.trace(REPACK_ERROR);
-    }
   }
   const tx = ethers6.Transaction.from(raw);
+  if (tx.isSigned() && (!signer || (await signer!.getAddress()) != tx.from!)) {
+    // us we are be unable to re-sign the encrypted tx, so must passthrough when
```
Comment (aefhm) [inline clients/js/src/compat.ts:579]:
```suggestion
    // we are be unable to re-sign the encrypted tx, so must passthrough when
```

File: contracts/test/eip155.ts [hunk 2]
```diff
@@ -36,8 +74,52 @@ describe('EIP-155', function () {
...
+    // Calldata should be encrypted when we wrap the wallet provider
+    let x = await provider.sendTransaction(signedTx);
+    expect(entropy(x.data)).gte(3.8);
+    expect(x.data).not.eq(calldata);
+
+    let r = await provider.waitForTransaction(x.hash);
+    expect(r.logs[0].data).equal(EXPECTED_EVENT);
   });
+
+  // TODO: test error conditions?
```
Comment (aefhm) [inline contracts/test/eip155.ts:124]: 👍

Comment (CedarMist) [reply contracts/test/eip155.ts:124]: This is being handled in another PR: https://github.com/oasisprotocol/sapphire-paratime

I'm going to leave the TODO: in for now

File: contracts/test/eip155.ts [hunk 3]
```diff
@@ -1,32 +1,73 @@
...
+  });
+
+  it('Wrapper encrypts self-signed transaction calldata', async function () {
+    const tx = await testContract.example();
+    expect(entropy(tx.data)).gte(EXPECTED_ENTROPY_ENCRYPTED);
+    expect(tx.data).not.eq(calldata);
+    expect(tx.data.length).eq(218);
+  });
+
+  /// Verifies that contract can sign a transaction, and we submit it via unwrapped provider
```
Comment (aefhm) [inline contracts/test/eip155.ts:57]:
```suggestion
  /// Verify that contracts can sign transactions for submission with an unwrapped provider
```

File: contracts/test/eip155.ts [hunk 4]
```diff
@@ -36,8 +77,53 @@ describe('EIP-155', function () {
...
       plainResp.hash,
     );
-    expect(receipt.logs[0].data).equal(
-      '0xfedcba9876543210fedcba9876543210fedcba9876543210fedcba9876543210',
-    );
+    expect(plainResp.data).eq(calldata);
+    expect(receipt.logs[0].data).equal(EXPECTED_EVENT);
+  });
+
+  /// Verifies that a contract can sign a transaction, and we can submit it via a wrapped provider
```
Comment (aefhm) [inline contracts/test/eip155.ts:84]:
```suggestion
  /// Verify that contracts can sign transactions for submission with a wrapped provider
```

File: contracts/test/eip155.ts [hunk 5]
```diff
@@ -36,8 +77,53 @@ describe('EIP-155', function () {
...
+  /// This lets transactions signed by other accounts pass-through the wrapped provider
```
Comment (aefhm) [inline contracts/test/eip155.ts:85]:
```suggestion
  /// Transactions signed by other accounts should pass-through the wrapped provider
```

File: contracts/test/eip155.ts [hunk 6]
```diff
@@ -36,8 +77,53 @@ describe('EIP-155', function () {
...
+      chainId: 0,
+    });
+
+    let plainResp = await provider.sendTransaction(signedTx);
+    let receipt = await provider.waitForTransaction(plainResp.hash);
+    expect(plainResp.data).eq(calldata);
+    expect(receipt.logs[0].data).equal(EXPECTED_EVENT);
+  });
+
+  /// Verifies that submitting a manually signed transaction will be encrypted
```
Comment (aefhm) [inline contracts/test/eip155.ts:106]:
```suggestion
  /// Verify that the wrapped wallet will encrypt a manually signed transaction
```

**Review by aefhm — APPROVED** [review]
Non-blocking comments.

**Pull request merged**